Sources/DDGSync/RecoveryPDFGenerator.swift

```diff
@@ -0,0 +1,191 @@
+//  limitations under the License.
+//
+
+import Foundation
+
+#if os(macOS)
+import AppKit
+#elseif os(iOS)
+import UIKit
+#endif
+
+protocol PDFGeneratorHelping {
+
+    associatedtype XFont
+    associatedtype XColor
+
+    var qrCodeYOffset: Int { get }
+    var textCodeYOffset: Int { get }
+    var textCodeFont: XFont { get }
+    var textCodeColor: XColor { get }
+
+    func pushGraphicsContext(_ context: CGContext)
+    func popGraphicsContext()
+    func flipContextIfNeeded(_ context: CGContext, boxHeight: CGFloat)
+    func drawTemplate(_ imageData: Data, in rect: CGRect)
+
+}
+
+public struct RecoveryPDFGenerator {
+
+    static let qrCodeSize = 175
+
+    let helper: any PDFGeneratorHelping
+
+    init(helper: any PDFGeneratorHelping) {
+        self.helper = helper
+    }
+
+    public init() {
+        self.init(helper: Helper())
+    }
+
+    public func generate(_ code: String) -> Data {
+        let data = NSMutableData()
+
+        let templateURL = Bundle.module.url(forResource: "SyncPDFTemplate", withExtension: "png")!
+        guard let templateData = try? Data(contentsOf: templateURL) else {
+            fatalError()
+        }
+
+        var mediaBox = CGRect(x: 0, y: 0, width: 612, height: 792)
+        let context = CGContext(consumer: CGDataConsumer(data: data)!, mediaBox: &mediaBox, nil)!
+
+        helper.pushGraphicsContext(context)
+        defer {
+            helper.popGraphicsContext()
+        }
+
+        // Prepare the PDF for drawing to
+        context.beginPDFPage(nil)
+        helper.flipContextIfNeeded(context, boxHeight: mediaBox.size.height)
+
+        // Draw the template image
+        helper.drawTemplate(templateData, in: CGRect(x: 0, y: 0, width: 612, height: 792))
+
+        // Draw the text
+        let paragraphStyle = NSMutableParagraphStyle()
+        paragraphStyle.lineHeightMultiple = 1.55
+        code.draw(in: CGRect(x: 290, y: helper.textCodeYOffset, width: 264, height: 1000), withAttributes: [
+                .font: helper.textCodeFont,
+                .foregroundColor: helper.textCodeColor,
+                .paragraphStyle: paragraphStyle,
+                .kern: 2
+            ])
+
+        // Draw the QRCode
+        let cgImage = qrcode(code, size: Self.qrCodeSize)
+        context.draw(cgImage, in: CGRect(x: 75, y: helper.qrCodeYOffset, width: Self.qrCodeSize, height: Self.qrCodeSize))
+
+        // Flush the data to the PDF file
+        context.endPDFPage()
+        context.closePDF()
+
+        return data as Data
+    }
+
+    func qrcode(_ text: String, size: Int) -> CGImage {
+        let data = Data(text.utf8)
+        let qrCodeFilter: CIFilter = CIFilter.init(name: "CIQRCodeGenerator")!
+        qrCodeFilter.setValue(data, forKey: "inputMessage")
+        qrCodeFilter.setValue("H", forKey: "inputCorrectionLevel")
+
+        guard let naturalSize = qrCodeFilter.outputImage?.extent.width else {
+            fatalError()
```

Comment:
Is this (and other calls to fatalError in this function) deliberate? I suppose that in the rare case when QR code generation fails, we could still put the code as text in the PDF and track it as error.

Reply:
It's deliberate yeah, but I'll add a task to do a pass over it when we productionise.  I didn't want to jump through any hoops to add pixels just now.